### PR TITLE
Generate: add warning when left padding should be used

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

As the title describes: add a warning when left padding should be used. Incorrect use of right padding is detected when:
1. the model is decoder-only;
2. there is a padding token in the last member of the sequence.

### CODE REVIEW

File: src/transformers/generation_flax_utils.py [hunk 1]
```diff
@@ -326,6 +326,13 @@ def generate(
         if decoder_start_token_id is None and self.config.is_encoder_decoder:
             raise ValueError("`decoder_start_token_id` has to be defined for encoder-decoder generation.")
 
+        # decoder-only models should use left-padding for generation
+        if not self.config.is_encoder_decoder and jnp.sum(input_ids[:, -1] == pad_token_id) > 0:
+            logger.warning(
+                "A decoder-only architecture is being used, but right-padding was detected! For correct generation "
+                "results, please set `padding_side='left'` when initializing the tokenizer."
+            )
```
Comment (gante) [inline src/transformers/generation_flax_utils.py:334]: I'm afraid this warning gets drown in the other logging/output the user might have.

Have we explored colored warnings for breaking behavior? (e.g. yellow)

**Review by sgugger — APPROVED** [review]
LGTM! I don't think we are sending that many warnings, so this should be clearly visible.

File: src/transformers/generation_utils.py [hunk 2]
```diff
@@ -1230,6 +1230,13 @@ def generate(
                 inputs_tensor, pad_token_id, eos_token_id
             )
 
+        # decoder-only models should use left-padding for generation
+        if not self.config.is_encoder_decoder and torch.sum(inputs_tensor[:, -1] == pad_token_id) > 0:
```
Comment (sgugger) [inline src/transformers/generation_utils.py:1234]: Looks like `inputs_tensor` is not always a tensor from the tests??

Comment (gante) [reply src/transformers/generation_utils.py]: it was not the cause -- when `pad_token_id` is `None`,  `inputs_tensor[:, -1] == pad_token_id` doesn't evaluate to a tensor :D

File: src/transformers/generation_flax_utils.py [hunk 3]
```diff
@@ -326,6 +326,14 @@ def generate(
         if decoder_start_token_id is None and self.config.is_encoder_decoder:
             raise ValueError("`decoder_start_token_id` has to be defined for encoder-decoder generation.")
 
+        # decoder-only models should use left-padding for generation (can't be checked with `trace=True`)
+        if not self.config.is_encoder_decoder and not trace:
```
Comment (gante) [inline src/transformers/generation_flax_utils.py:330]: @patrickvonplaten -- without the `and not trace`, a `jax._src.errors.ConcretizationTypeError` is raised, because it's a boolean that depends on the contents of `input_ids`. This seemed the wisest path to inform the users without breaking the code

Also, FLAX is more picky than XLA TF, which doesn't complain about this check :D

Comment (patrickvonplaten) [reply src/transformers/generation_flax_utils.py:330]: Sounds good!

**Review by patrickvonplaten — APPROVED** [review]
